awesome_cart/awc.py

```diff
@@ -1652,6 +1678,133 @@ def cart(data=None, action=None):
+			"net_total": quotation.base_net_total,
+			"is_ground_shipping": True if "GROUND" in shipping_method else False
+		})
+
+		if has_total_limit or \
+			(discount == 0 and ( \
+				len(frappe.response.get("coupon_insert_items", [])) == 0 and \
+				not has_services
+			)):
+
+			remove_items = find_items_with_insert_id(awc["items"])
+			remove_success, removed_quotation_item_ids, awc_items = remove_from_cart(
+				remove_items, awc["items"])
+
+			if remove_success:
+				# replace items with filtered version
+				awc["items"] = awc_items
+				removed_ids = removed_ids + list(set(removed_quotation_item_ids) - set(removed_ids))
+
+			is_valid = False
+			success = False
+			awc["discounts"] = None
+			quotation.coupon_code = ""
+			if not has_total_limit:
+				msg = "Coupon is invalid for the current cart."
+		else:
+			awc["discounts"] = coupon_state
+
+			insert_items = frappe.response.get("coupon_insert_items", [])
+
+			add_items = []
+			remove_items = []
+
+			for x in insert_items:
+				insert_logic = x.get("options", {}).get("insert_logic", False)
+				if insert_logic:
+					is_truthy = zscript.run(insert_logic, icontext)
+				else:
+					is_truthy = False
+
+				if is_truthy:
+					if not find_item_by_insert_id(awc["items"], x.get("options").get("insert_id")):
+						add_items.append(x)
+				else:
+					remove_insert_id = find_item_by_insert_id(awc["items"], x.get("options").get("insert_id"))
+					if remove_insert_id:
+						remove_items.append(remove_insert_id)
+
+			quotation.coupon_code = coupon
+			quotation_is_dirty = True
+
+			# add passing logic items
+			if len(add_items) > 0:
+				quotation_is_dirty, removed_ids = add_to_cart(add_items, awc_session, quotation, True)
+
+			# remove non passing logic items
+			if len(remove_items) > 0:
+				remove_success, removed_quotation_item_ids, awc_items = remove_from_cart(
+					remove_items, awc["items"])
+				if remove_success:
+					# replace items with filtered version
+					awc["items"] = awc_items
+					removed_ids += list(set(removed_quotation_item_ids) - set(removed_ids))
+
+			success = True
+	else:
+		success = False
+		awc["discounts"] = None
+		quotation.coupon_code = None
+		if "coupon" in awc["totals"]:
+			del awc["totals"]["coupon"]
+
+		remove_items = find_items_with_insert_id(awc["items"])
+		remove_success, removed_quotation_item_ids, awc_items = remove_from_cart(
+			remove_items, awc["items"])
+
+		if remove_success:
+			# replace items with filtered version
+			awc["items"] = awc_items
+			removed_ids += list(set(removed_quotation_item_ids) - set(removed_ids))
+
+	if len(removed_ids) > 0:
+		quotation_items = [ \
+			itm for itm in quotation.get("items", []) \
+									if itm.name not in removed_ids \
+		]
+		awc["items"] = [ i for i in awc.get("items", []) if i.get("id") not in removed_ids ]
+		quotation.set("items", quotation_items)
+
+	return (success, msg, quotation_is_dirty, is_valid, removed_ids)
+
+def find_item_by_insert_id(awc_items, id):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'id'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)